File: ast/src/testing/csharp/MethodChainFactory.cs
// Tests static-factory call resolution via the uppercase-identifier fallback.
//
// WidgetFactory.Create() is a static method returning Widget.  The resolver
// must recognise "WidgetFactory" as a class-name receiver (not a local var)
// so it can look up the method in the graph.  Then w.Render() resolves
// Widget::Render via find_method_in_class.
//
// @ast node: Class "Widget"
// @ast node: Class "WidgetClient"
// @ast node: Class "WidgetFactory"
// @ast node: Function "Create"
// @ast node: Function "Render"
// @ast node: Function "WidgetClient"
// @ast node: Function "Run"
// @ast edge: Calls -> Function "Create" "MethodChainFactory.cs"
// @ast edge: Calls -> Function "Render" "MethodChainFactory.cs"
namespace MethodChain;

class Widget
{
    public void Render() { }
}

class WidgetFactory
{
    public static Widget Create() { return new Widget(); }
}

class WidgetClient
{
    public WidgetClient() { }

    public void Run()
    {
        Widget w = WidgetFactory.Create();
        w.Render();
    }
}

File: ast/src/testing/java/src/main/java/graph/stakgraph/java/nonweb/MethodChainFactory.java
// Tests static-factory call resolution.
//
// WidgetFactory.create() is a static method returning Widget.  The resolver
// must recognise "WidgetFactory" as a class-name receiver (not a local var)
// so it can look up (WidgetFactory, create) in method_returns → Widget.
// Then w.render() resolves Widget::render via class_fields dispatch.
//
// @ast node: Class "Widget"
// @ast node: Function "render"
// @ast node: Class "WidgetFactory"
// @ast node: Function "create"
// @ast node: Class "WidgetClient"
// @ast node: Function "WidgetClient"
// @ast node: Function "run"
// @ast edge: Calls -> Function "create" "MethodChainFactory.java"
// @ast edge: Calls -> Function "render" "MethodChainFactory.java"
// @ast node: Instance "w"
// @ast edge: Of -> Class "Widget" "MethodChainFactory.java"
// @ast node: Var "w"
package graph.stakgraph.java.nonweb;

class Widget {
    public void render() {}
}

class WidgetFactory {
    public static Widget create() {
        return new Widget();
    }
}

class WidgetClient {
    public WidgetClient() {}

    public void run() {
        Widget w = WidgetFactory.create();
        w.render();
    }
}
